src/main/java/com/odysseusinc/athena/service/impl/NotificationServiceImpl.java

```diff
@@ -0,0 +1,114 @@
+import com.odysseusinc.athena.repositories.athena.NotificationRepository;
+import com.odysseusinc.athena.repositories.athena.VocabularyConversionRepository;
+import com.odysseusinc.athena.repositories.v5.VocabularyRepository;
+import com.odysseusinc.athena.service.NotificationService;
+import com.odysseusinc.athena.service.mail.EmailService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.commons.lang.StringUtils.equalsIgnoreCase;
+
+@Service
+@Transactional
+public class NotificationServiceImpl implements NotificationService {
+
+    private final EmailService emailService;
+    private final NotificationRepository notificationRepository;
+    private final UserService userService;
+    private final VocabularyConversionRepository vocabularyConversionRepository;
+    private final VocabularyRepository vocabularyRepository;
+
+    @Autowired
+    public NotificationServiceImpl(EmailService emailService, NotificationRepository notificationRepository, UserService userService, VocabularyConversionRepository vocabularyConversionRepository, VocabularyRepository vocabularyRepository) {
+        this.emailService = emailService;
+        this.notificationRepository = notificationRepository;
+        this.userService = userService;
+        this.vocabularyConversionRepository = vocabularyConversionRepository;
+        this.vocabularyRepository = vocabularyRepository;
+    }
+
+    @Override
+    public void createNotificationSubscriptions(Long userId, String[] vocabularyCodes) {
+
+        for (String vocabularyCode : vocabularyCodes) {
+            Notification current = notificationRepository.findByUserIdAndVocabularyCode(userId, vocabularyCode);
+            if (current == null) {
+                final VocabularyV5 vocabulary = vocabularyRepository.findOne(vocabularyCode);
+                if (vocabulary != null) {
+                    final VocabularyConversion vocabularyConversion = vocabularyConversionRepository.findByIdV5(vocabularyCode);
+                    final Notification newNotification = new Notification(userId, vocabularyConversion, vocabularyCode, vocabulary.getVersion());
+                    notificationRepository.save(newNotification);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void deleteNotificationSubscription(Long userId, String vocabularyCode) {
+        Notification current = notificationRepository.findByUserIdAndVocabularyCode(userId, vocabularyCode);
+        if (current != null) {
+            notificationRepository.delete(current);
+        }
+    }
+
+    /**
+     * this method is required for the transition of already existing subscriptions
+     */
+    @Override
+    public void ensureVocabularyVersionAndCodeAreSet() {
```

Comment:
layer bounds: method is completely internals of service, don't make it public

Reply:
it is temporary method which is required to migrate current notification records without version

I cannot put it in flywaydb because it uses data from different db (vocabulary_v5)
It will be removed in next release